eZ/Publish/Core/Persistence/Legacy/Content/Type/ContentUpdater/Action/AddField.php

```diff
@@ -77,43 +77,66 @@ public function __construct(
     /**
      * Applies the action to the given $content
      *
-     * @param \eZ\Publish\SPI\Persistence\Content\ContentInfo $contentInfo
+     * @param int $contentId
      */
-    public function apply( ContentInfo $contentInfo )
+    public function apply( $contentId )
     {
-        $languageCodeSet = array();
-        $versionNumbers = $this->contentGateway->listVersionNumbers( $contentInfo->id );
+        $versionNumbers = $this->contentGateway->listVersionNumbers( $contentId );
+        $languageCodeToFieldId = array();
 
-        $contentRows = $this->contentGateway->load( $contentInfo->id, $contentInfo->currentVersionNo );
-        $contentList = $this->contentMapper->extractContentFromRows( $contentRows );
-        $content = $contentList[0];
-
-        foreach ( $content->fields as $field )
+        foreach ( $versionNumbers as $versionNo )
         {
-            if ( isset( $languageCodeSet[$field->languageCode] ) )
-            {
-                continue;
-            }
+            $contentRows = $this->contentGateway->load( $contentId, $versionNo );
+            $contentList = $this->contentMapper->extractContentFromRows( $contentRows );
+            $content = $contentList[0];
+            $languageCodeSet = array();
 
-            $languageCodeSet[$field->languageCode] = true;
-
-            foreach ( $versionNumbers as $versionNo )
+            // Each subsequent Content version can have additional language(s)
+            foreach ( $content->fields as $field )
             {
-                $this->insertField(
+                $languageCode = $field->languageCode;
+
+                // Add once for each language per version
+                if ( isset( $languageCodeSet[$languageCode] ) )
+                {
+                    continue;
+                }
+
+                $languageCodeSet[$languageCode] = true;
+
+                // Check if field was already inserted for current language code,
+                // in that case we need to preserve its ID across versions
+                if ( isset( $languageCodeToFieldId[$languageCode] ) )
```

Comment:
True, but not without the middle part; I don't like to use it if it makes code less readable.